### PR TITLE
fix: document openapi-core package usage + slightly improve the API

### PR DESCRIPTION
**Pull request by RomanHotsiy**
## What/Why/How?

Folks are using our openapi-core package but it is not documented so they are stuck or end up using some internals or legacy functionality which is not intended for external use (for example: https://github.com/Redocly/redocly-cli/issues/1274).

This PR improves the public API slightly and also adds documentation the most common use cases so people can rely on the documented functionality.

This PR also adds an explicit warning to avoid using anything that is not documented. This will save time for both them and us.

We may want to move this documentation to our marketing site but I would prefer to keep it in the readme and to link here from main readme/marketing site.

## Reference

Related to https://github.com/Redocly/redocly-cli/issues/1274

## Testing

We need to cover all of the examples from basic usage section by tests. @tatomyr could you help me with that?

## Check yourself

- [x] Code is linted
- [ ] Tested with redoc/reference-docs/workflows (internal)
- [ ] All new/updated code is covered with tests

## Security

- [x] Security impact of change has been considered
- [x] Code follows company security practices and guidelines

### CODE REVIEW

**Review by lornajane — CHANGES_REQUESTED** [review]
Thank you! This is golden, and exactly what I was hoping for! Marking as needing changes as I think all code samples need words with them before we publish, but I can work on those.

**Comment by lornajane** [comment]
Should I be able to run these code snippets? Tried the first one to get a sense of what the user gets back in `lintResults` but it tells me:
```
Promise { <pending> }
/home/lorna/temp/node_modules/@redocly/openapi-core/lib/lint.js:44
        const rules = config.getRulesForOasVersion(specMajorVersion);
                             ^

TypeError: Cannot read properties of undefined (reading 'getRulesForOasVersion')
```

What am I missing? I'm using a `redocly.yaml` file with a single rule in it, linting with CLI doesn't report config lint problems.

File: packages/core/README.md [hunk 1]
```diff
@@ -2,24 +2,250 @@
...
-### Lint
+### Lint from file system
 
 ```js
-import { formatProblems, lint, loadConfig } from '@redocly/openapi-core';
+import { lint, loadConfig } from '@redocly/openapi-core';
 
 const pathToApi = 'openapi.yaml';
-const config = loadConfig({ configPath: 'optional/path/to/.redocly.yaml' });
+const config = loadConfig({ configPath: 'optional/path/to/redocly.yaml' });
```
Comment (RomanHotsiy) [inline packages/core/README.md:19]:
```suggestion
const config = await loadConfig({ configPath: 'optional/path/to/redocly.yaml' });
```

File: packages/core/README.md [hunk 2]
```diff
@@ -2,24 +2,250 @@
...
-### Bundle
+### Bundle from file system
 
 ```js
-import { formatProblems, bundle, loadConfig } from '@redocly/openapi-core';
+import { bundle, loadConfig } from '@redocly/openapi-core';
 
 const pathToApi = 'openapi.yaml';
-const config = loadConfig({ configPath: 'optional/path/to/.redocly.yaml' });
+const config = loadConfig({ configPath: 'optional/path/to/redocly.yaml' });
```
Comment (RomanHotsiy) [inline packages/core/README.md:29]:
```suggestion
const config = await loadConfig({ configPath: 'optional/path/to/redocly.yaml' });
```

File: packages/core/README.md [hunk 3]
```diff
@@ -2,24 +2,250 @@
...
+
+const CustomRule = (ruleOptions) => {
+  return {
+    Operation() {
+      // some rule logic
+    },
+  };
+};
+
+const config = createConfig({
```
Comment (RomanHotsiy) [inline packages/core/README.md:72]:
```suggestion
const config = await createConfig({
```

**Comment by RomanHotsiy** [comment]
I did not test all of those examples myself. The await was missing I think. I just fixed it above ☝️

File: packages/core/README.md [hunk 4]
```diff
@@ -2,24 +2,267 @@
...
+import { bundle, loadConfig } from '@redocly/openapi-core';
 
 const pathToApi = 'openapi.yaml';
-const config = loadConfig({ configPath: 'optional/path/to/.redocly.yaml' });
-const { bundle, problems } = await bundle({ ref: pathToApi, config });
+const config = await loadConfig({ configPath: 'optional/path/to/redocly.yaml' });
+const bundleResults = await bundle({ ref: pathToApi, config });
+```
+
+In `bundleResults`, the `bundle.parsed` field has the bundled API description.
```
Comment (lornajane) [inline packages/core/README.md:41]: @RomanHotsiy is this good advice? I improvised, please check!

File: packages/core/README.md [hunk 5]
```diff
@@ -2,24 +2,267 @@
...
 
 ```js
-import { formatProblems, bundle, loadConfig } from '@redocly/openapi-core';
+import { bundle, loadConfig } from '@redocly/openapi-core';
 
 const pathToApi = 'openapi.yaml';
-const config = loadConfig({ configPath: 'optional/path/to/.redocly.yaml' });
-const { bundle, problems } = await bundle({ ref: pathToApi, config });
+const config = await loadConfig({ configPath: 'optional/path/to/redocly.yaml' });
+const bundleResults = await bundle({ ref: pathToApi, config });
```
Comment (lornajane) [inline packages/core/README.md:38]: We're going to stop linting, and I couldn't reuse the `bundle` variable name so I switched the output here.

**Review by lornajane — APPROVED** [review]

**Comment by lornajane** [comment]
@tatomyr do we need to do anything in particular to make this README be used as the npm README? I have a vague memory of tripping over this in the past.

**Comment by tatomyr** [comment]
@lornajane it is already: https://www.npmjs.com/package/@redocly/openapi-core.

File: .changeset/two-shoes-sniff.md [hunk 6]
```diff
@@ -0,0 +1,5 @@
+---
+'@redocly/openapi-core': patch
+---
+
+fix: enhance public API by small typescript typing fix and exporting new function `bundleFromString`
```
Comment (tatomyr) [inline .changeset/two-shoes-sniff.md:5]: Correcting this to fit the changelog format:

```suggestion
Enhanced public API by small typescript typing fix and exporting new function `bundleFromString`.
```

**Comment by lornajane** [comment]
What's left before we merge this? I approved, there's some changelog housekeeping and one thing I need someone to check it makes sense. Anything else? Can we make this happen?

**Comment by tatomyr** [comment]
@lornajane I'm trying to find some time to write a couple of tests 😅. Will try to finish as soon as possible.

**Comment by RomanHotsiy** [comment]
I'm okay to merge it whenever @tatomyr is ready.

File: packages/core/src/__tests__/bundle.test.ts [hunk 7]
```diff
@@ -261,3 +257,21 @@ describe('bundle', () => {
...
+      bundle: { parsed, ...rest },
+      problems,
+    } = await bundleFromString({
+      config: await createConfig(`
+        extends:
+        - recommended
+      `),
+      source: testDocument.source.body,
+    });
+    console.log(rest);
```
Comment (SmoliyY) [inline packages/core/src/__tests__/bundle.test.ts:273]: I think you forgot to remove this

**Review by tatomyr — APPROVED** [review]

File: packages/core/src/__tests__/bundle.test.ts [hunk 8]
```diff
@@ -233,4 +232,46 @@ describe('bundle', () => {
...
```
Comment (tatomyr) [inline packages/core/src/__tests__/bundle.test.ts:273]:
```suggestion
```

**Review by SmoliyY — APPROVED** [review]

File: packages/core/src/__tests__/bundle.test.ts (hunk 7 above)
Comment (tatomyr) [reply packages/core/src/__tests__/bundle.test.ts]: Indeed, thanks!

**Comment by tatomyr** [comment]
I added the missing tests. @RomanHotsiy you may proceed further.

**Comment by RomanHotsiy** [comment]
> there's some changelog housekeeping and one thing I need someone to check it makes sense

@lornajane do you want to adjust changelog and what is the one thing? Or should I merge and you'll do it in a separate PR?

**Comment by tatomyr** [comment]
I've already adjusted the changeset.

**Comment by lornajane** [comment]
@RomanHotsiy this one thing: https://github.com/Redocly/redocly-cli/pull/1284#discussion_r1348571746 - I think `bundle.parsed` is where the user can find the bundled API description, is that right? I tried it and it seemed like this would be the right field to use.

File: packages/core/README.md (hunk 4 above)
Comment (RomanHotsiy) [reply packages/core/README.md:41]: Yes

**Comment by RomanHotsiy** [comment]
@lornajane yes, this is correct. I'm going ahead and merging the PR.

Please, cut a new release when you are ready.

**Pull request merged**